### PR TITLE
Remove the need to use Erlang/Elixir's startup scripts

### PR DESCRIPTION
**Pull request by doawoo**
We can now launch the app directly from Zig

### CODE REVIEW

**Comment by doawoo** [comment]
@bsedat @QuinnWilton I believe we can remove the base64 encoded hack for windows, since we're no longer going to be passing the args through any of the batch files, and going straight to `erl.exe`

File: src/erlang_launcher.zig [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    const erts_version_name = try std.fmt.allocPrint(allocator, "erts-{s}", .{ meta.erts_version });
+    var erl_bin_path = try fs.path.join(allocator, &[_][]const u8{ install_dir, erts_version_name, "bin", get_erl_exe_name() });
+
+    // Read the Erlang COOKIE file for the release
+    const release_cookie_file = try fs.openFileAbsolute(release_cookie_path, .{ .read = true, .write = false });
+    const release_cookie_content = try release_cookie_file.readToEndAlloc(allocator, MAX_READ_SIZE);
+
+    // Set all the required release environment variables
+
+    try env_map.put("NODE_NAME", meta.app_name[0..]);
```
Comment (bsedat) [inline src/erlang_launcher.zig:42]: Is this used by Erlang? I don't see any references to it in the erl documentation.

**Comment by bsedat** [comment]
> @bsedat @QuinnWilton I believe we can remove the base64 encoded hack for windows, since we're no longer going to be passing the args through any of the batch files, and going straight to `erl.exe`

Nice - I was going to do some investigation since I was seeing some quotes getting stripped from argv on Linux as well but if we go right to erl.exe that may not be necessary.

**Comment by bsedat** [comment]
I wonder if we should have apps set `include_erts: false` and `include_executables_for: []` in their mix release?

**Comment by bsedat** [comment]
I think we'll also need to pass along the `-config` arg to erlang. I'm still looking for where the runtime config provider receives the temporary config file to give to Erlang.

How should we handle `env.sh`/`env.bat`? We may want to add compatibility with some of the existing release configuration environmen vars (https://hexdocs.pm/mix/1.13/Mix.Tasks.Release.html#module-environment-variables), although I am a bit hesitant about running environment-specific shell files again. Perhaps it is something we allow to be configured via a Zig plugin?

**Comment by bsedat** [comment]
> I think we'll also need to pass along the `-config` arg to erlang. I'm still looking for where the runtime config provider receives the temporary config file to give to Erlang, but it looks like runtime_config is set as a comment in the `sys.config` and only set to true if the system is configured to reboot after configuration is generated.
> 
> How should we handle `env.sh`/`env.bat`? We may want to add compatibility with some of the existing release configuration environmen vars (https://hexdocs.pm/mix/1.13/Mix.Tasks.Release.html#module-environment-variables), although I am a bit hesitant about running environment-specific shell files again. Perhaps it is something we allow to be configured via a Zig plugin?

Looks like the runtime config temporary file is opt-in now since Elixir 1.11: https://github.com/elixir-lang/elixir/commit/bc5bb85a1763f2e626b77ea67576e116906eee20. Perhaps we should just call it out as not supported for Burrito and instead just pass along the static `sys.config`.

File: src/erlang_launcher.zig (hunk 1 above)
Comment (doawoo) [reply src/erlang_launcher.zig:42]: Ah yup this was left over from me tracing the env the stock startup scripts did, I'll remove that!

File: src/erlang_launcher.zig [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+const builtin = @import("builtin");
+const fs = std.fs;
+const log = std.log;
+const metadata = @import("metadata.zig");
+const win_asni = @cImport(@cInclude("win_ansi_fix.h"));
+
+const MetaStruct = metadata.MetaStruct;
+const BufMap = std.BufMap;
+
+const MAX_READ_SIZE = 1000000000;
```
[the remark this answers is not in the document]
Comment (doawoo) [reply src/erlang_launcher.zig:12]: According to the OTP source, that's exactly right:
https://github.com/erlang/otp/blob/7e332d22c4cd2a1da9f7207c4e978d3c5ccb944e/lib/erl_interface/include/ei.h#L226

I'll revise :)

**Comment by doawoo** [comment]
> I wonder if we should have apps set `include_erts: false` and `include_executables_for: []` in their mix release?

this is a good note, I want to do some experiments on that...

**Comment by doawoo** [comment]
> > I think we'll also need to pass along the `-config` arg to erlang. I'm still looking for where the runtime config provider receives the temporary config file to give to Erlang, but it looks like runtime_config is set as a comment in the `sys.config` and only set to true if the system is configured to reboot after configuration is generated.
> > How should we handle `env.sh`/`env.bat`? We may want to add compatibility with some of the existing release configuration environmen vars (https://hexdocs.pm/mix/1.13/Mix.Tasks.Release.html#module-environment-variables), although I am a bit hesitant about running environment-specific shell files again. Perhaps it is something we allow to be configured via a Zig plugin?
> 
> Looks like the runtime config temporary file is opt-in now since Elixir 1.11: [elixir-lang/elixir@bc5bb85](https://github.com/elixir-lang/elixir/commit/bc5bb85a1763f2e626b77ea67576e116906eee20). Perhaps we should just call it out as not supported for Burrito and instead just pass along the static `sys.config`.

I'll add some signage about Burrito only running Erlang in embedded mode, which as specified by the manual will only look at sys.config

**Comment by bsedat** [comment]
> > > I think we'll also need to pass along the `-config` arg to erlang. I'm still looking for where the runtime config provider receives the temporary config file to give to Erlang, but it looks like runtime_config is set as a comment in the `sys.config` and only set to true if the system is configured to reboot after configuration is generated.
> > > How should we handle `env.sh`/`env.bat`? We may want to add compatibility with some of the existing release configuration environmen vars (https://hexdocs.pm/mix/1.13/Mix.Tasks.Release.html#module-environment-variables), although I am a bit hesitant about running environment-specific shell files again. Perhaps it is something we allow to be configured via a Zig plugin?
> > 
> > 
> > Looks like the runtime config temporary file is opt-in now since Elixir 1.11: [elixir-lang/elixir@bc5bb85](https://github.com/elixir-lang/elixir/commit/bc5bb85a1763f2e626b77ea67576e116906eee20). Perhaps we should just call it out as not supported for Burrito and instead just pass along the static `sys.config`.
> 
> I'll add some signage about Burrito only running Erlang in embedded mode, which as specified by the manual will only look at sys.config

What do you think about Burrito checking the mix release configuration to see if `reboot_system_after_config` is set to true and erroring since we won't properly pull in the generated config?

File: lib/util/args.ex [hunk 3]
```diff
@@ -1,37 +1,11 @@
...
-    end
-  end
-
-  defp is_windows? do
-    match?({:win32, _}, :os.type())
-  end
-
-  defp args_are_encoded? do
-    System.get_env("_ARGUMENTS_ENCODED", "0") == "1"
+    :init.get_plain_arguments() |> Enum.map(&to_string/1)
```
Comment (bsedat) [inline lib/util/args.ex:9]: I'm not sure why it changed but `Args.get_arguments/0` is returning a list with a single string in it now.

**Comment by bsedat** [comment]
Tried it with Windows and it seems to be working, so just have a final question around the arguments and then I'd be ready to merge!

File: lib/util/args.ex (hunk 3 above)
Comment (doawoo) [reply lib/util/args.ex:9]: This should be resolved in the latest commit I just pushed!

**Pull request merged**